### DISCUSSION TITLE
Hybrid property in the classical mapping

### DISCUSSION THREAD

**Discussion by Tserewara** [sqlalchemy/sqlalchemy#6036] [Usage Questions] [answered]
First of all, I really appreciate your efforts to provide such a great tool. I'd like to know if it's possible to accomplish the following.

I have the following classes:

```py
    class Password:
    
        def __init__(self, encryptor: AbstractEncryptor, value: str = None):
            self._encryptor = encryptor
            self._value = value
    
        def __repr__(self):
            return '<Password object {}>'.format(self._value)
    
        def __eq__(self, value: str) -> bool:
            return self._encryptor.verify(value, self._value)
    
        @property
        def value(self) -> str:
            return self._value
    
        @classmethod
        def validate_strength(cls, value: str) -> (bool, dict):
            pass # logic to validate goes here
  
        @value.setter
        def value(self, value: str):
            valid_password, _ = self.validate_strength(value)
            if not valid_password:
                raise exceptions.PasswordStrengthError(
                    'A strong password should contain at least 8 characters, '
                    '1 digit, 1 symbol, 1 uppercase letter, and 1 lowercase '
                    'letter'
                )
            self._value = self._encryptor.encrypt(value)
```
and

```py
    class Credential:
    
        def __init__(
                self,
                username: str,
                password: str = None,
                encryptor: AbstractEncryptor = Encryptor,
                active: bool = True
        ):
            self.username = username
            self._password = Password(encryptor, password)
            self._active = active
    
        @property
        def password(self):
            return self._password
    
        @password.setter
        def password(self, value: str):
            self._password.value = value
    
        def __eq__(self, other: Credential) -> bool:
            return (self.username == other.username and
                    self._password.value == other.password)
    
        def verify_password(self, value: str) -> bool:
            return self._password == value
    
        def deactivate(self):
            self.active = False
```
I'm trying to map ```Credential``` using the classical mapping in SqlAlchemy.

```py
    credentials = Table('credentials', metadata,
                        Column('id', Integer, primary_key=True,
                               autoincrement=True),
                        Column('username', String, index=True,
                               unique=True, nullable=False),
                        Column('password', String, nullable=False),
                        Column('active', Boolean)
                        )

    mapper(credential.Credential, credentials, properties={
        'password': synonym('_password', map_column=True,
                            descriptor=credential.Credential.password)
    })
```
But the problem is I can't make SqlAlchemy see the actual calculated string value of password as I defined in my getter. Then, when I try to run it, it raises InterfaceError since it tries to save a `Password` object instead of a string. I need to make it save the calculated value of `Password`. 

PS: I know it's possible to make it with `hybrid_property` but it only seems to be available in the declarative mapping. Is there a way to do in classical mapping?

**Versions**
 - OS: Ubuntu 20.04
 - Python: 3.8
 - SQLAlchemy: 1.3.23
 - Database: PostgreSQL13.1
 - DBAPI: psycopg2

**Comment by zzzeek** [comment, marked as answer]
So the best example of this kind of thing is in my old blog post at https://techspot.zzzeek.org/2011/10/21/hybrids-and-value-agnostic-types/ , and it then continues on to show the same thing in terms of the TypeDecorator class at https://techspot.zzzeek.org/2011/10/29/value-agnostic-types-part-ii/ .       That is, instead of mapping "password" to String, map it to a custom type called PasswordType.   then it will work in any situation without the need for hybrids.

hybrids themselves are nothing more than [Python descriptors](https://docs.python.org/3/howto/descriptor.html) and whether you use declarative or classical mapping or even use the ORM at all, hybrids are avialble just by placing them on your class.

Beyond all that it seems like you are only dealing with instance-level access in which case the `@property` you have on your Credential class should be sufficient for converting between Password and string.  I think you just have it backwards, "Password" is the public facing object and "_password" is the string version.

```python

class Credential:

    def __init__(
            self,
            username: str,
            password: str = None,
            encryptor: AbstractEncryptor = Encryptor,
            active: bool = True
    ):
        self.username = username
        self._password = password
        self._encryptor = encryptor
        self._active = active

    @property
    def password(self):
        return Password(self._encryptor, self._password)

    @password.setter
    def password(self, value: Password):
        self._password = value._value
        self._encryptor = value.encryptor

```

The way you have it now, there's an asymmetry between your get and set; "get" is retunring a Password object but "set" is accepting a string.  you'd want those to both work with the same object type.

**Comment by Tserewara** [comment]
Thank you so much for taking time to reply. I'm studying the material you suggested. 

I might be missing something, but shouldn't SQLAlchemy be able to get the value of `Password` out of the box since I set it in the class?

```py
class Password:
 
 @property
    def value(self) -> str:
        return self._value # the string value of password is returned
        
class Credential:

    def __init__(
            self,
            username: str,
            password: str = None,
            encryptor: AbstractEncryptor = Encryptor,
            active: bool = True
    ):
        self.username = username
        self._password = Password(encryptor, password)
        self._active = active

    @property
    def password(self):
        return self._password.value # it access the string value of Password. Shouldn't SQLAlchemy get the calculated value using this?
```
By the way, I just realized I forget to add `return self._password.value` but I actually did it in my code.

I'm able to access its value when I instantiate `Credential`.

```py
credential = Credential("John")
credential.password = "A@Password999"

print(credential.password) # returns the hashed password as it is meant

```
But when I run my test it raises an InterfaceError since it tries to save the object `Password` instead of its calculated value from the property. 

```
E       sqlalchemy.exc.InterfaceError: (sqlite3.InterfaceError) Error binding parameter 1 - probably unsupported type.
E       [SQL: INSERT INTO credentials (username, password) VALUES (?, ?)]
E       [parameters: ('tserewara', <src.articles.domain.entities.credential.SimplePassword object at 0x000001B37C638340>)]
E       (Background on this error at: http://sqlalche.me/e/13/rvf5)
```
 Shouldn't SQLAlchemy get the calculated value using the property I set?

**Reply by zzzeek** [reply]
> 
> I might be missing something, but shouldn't SQLAlchemy be able to get the value of `Password` out of the box since I set it in the class?

I'm not following your terminology, what is "the box" ?   


> 
> ```python
> class Password:
>  
>  @property
>     def value(self) -> str:
>         return self._value # the string value of password is returned
>         
> class Credential:
> 
>     def __init__(
>             self,
>             username: str,
>             password: str = None,
>             encryptor: AbstractEncryptor = Encryptor,
>             active: bool = True
>     ):
>         self.username = username
>         self._password = Password(encryptor, password)
>         self._active = active
> 
>     @property
>     def password(self):
>         return self._password.value # it access the string value of Password. Shouldn't SQLAlchemy get the calculated value using this?

when someone accesses Credential(...).password, would you like it to return a string or a Password object?  your code example is ambiguous in this regard.

SQLAlchemy itself doesn't "get" anything except for what you've mapped to the "Column("password", String)" column, which in your example is the attribute named "_password".  that value has to be a string at all times as long as the Column() is using String as the datatype.


> ```
> 
> By the way, I just realized I forget to add `return self._password.value` but I actually did it in my code.
> 
> I'm able to access its value when I instantiate `Credential`.
> 
> ```python
> credential = Credential("John")
> credential.password = "A@Password999"
> 
> print(credential.password) # returns the hashed password as it is meant
> ```
> 
> But when I run my test it raises an InterfaceError since it tries to save the object `Password` instead of its calculated value from the property.
> 
> ```
> E       sqlalchemy.exc.InterfaceError: (sqlite3.InterfaceError) Error binding parameter 1 - probably unsupported type.
> E       [SQL: INSERT INTO credentials (username, password) VALUES (?, ?)]
> E       [parameters: ('tserewara', <src.articles.domain.entities.credential.SimplePassword object at 0x000001B37C638340>)]
> E       (Background on this error at: http://sqlalche.me/e/13/rvf5)
> ```
> 
> Shouldn't SQLAlchemy get the calculated value using the property I set?

**Reply by Tserewara** [reply]
Hmm I think I got it wrong, sorry. I thought I could map the calculated attribute directly to SQLAlchemy so it would save it as string. I would like to save the string value provided by Credential.password. Sorry again for my lack of knowledge. I really appreciate your patience.

**Reply by Tserewara** [reply]
I just fixed my poor class, made it return the value I wanted as a strign and got it running. Thank you so much for your guidance and for SQLAlchemy. I couldn't fix my problem without your help. 😄😄
